### ISSUE TITLE
Law Question/Confustion - Action in district after district removal?

### ISSUE THREAD

**Issue by Deantwo** [StrangeLoopGames/EcoIssues#8841]
Law question/confusion. Maybe you can answer this @MarjieAVolk.

When we use the `action place item where (in district test1)`, at what point is the action considered to be in the district? On time of the action or the time of the law check?

For example if i make a periodic law that tax people for the amount of store objects they have in a district (placed - picked up), are stores that were placed before the district was defined counted? What if the district is resized and now some stores are outside the district, are they still counted?

I guess the question borrows down to, is the district it is preformed in saved in the action? Or is the location saved and then the location checked if it is inside a district upon query?

Same questions likely work for "on own property" and similar where conditions.

**Comment by MarjieAVolk** [comment]
Whether each action took place in a district is based on the situation at the time of the action.  As you say, the district and/or deed is saved, and not the actual position.  So you're right, if you define a district and start taxing based on stores placed, stores that had already been there will not be counted.

If I were to add some values that determine present numbers of owned objects, etc, I don't believe it would be possible currently to do that through stats (that is to say, the methods you already have of using stats are basically all stats can do in that regard currently).  I think the two main options then are either: calculate that stuff directly from the game state; or provide two options for 'in district'/'on property' type queries - one for the district at the time of the action, and one for the district now based on the position of the action.

**Comment by Deantwo** [comment]
Yeah I figured the action stats were stored like that. Adding two separate options, one for "time of action" and one for "current location status" would be useful. Not totally error-proof, but better.

Can the action also maybe have a "owner at time of action" and "current owner"? Or is a reference to the object not saved with the stat action?

This whole thing was most of all a theory I had, so both happy and sad that I was correct.

**Comment by MarjieAVolk** [comment]
Okay, well I will think about it more.  I'm not sure how to word those 2 options clearly.

For world objects, the actual object is tracked in the action.  But if the object has been picked up from the world, it will no longer be able to find it, so it won't necessarily have a 'current owner'.  I'm not sure what kind of laws you would want to make with this, or how it would fit in.

**Comment by D3nnis3n** [comment]
I agree with Deantwo. It would be nice to have options to define if you want the actions to be counted based on their location or based on the time (e.g. when the district has been created).

**Comment by Deantwo** [comment]
Counting everything in actions just has so many possibilities for errors. But I do understand that checking the world database every time a law is triggered could cause issues.
So not sure about the best way to fix any of this.

Especially admin intervention will ruin a lot of the action counting laws.

**Comment by WeaselDog** [comment]
@IcedForge Can you check if this is still an issue?

**Comment by IcedForge** [comment]
Not entirely sure what would be the issue?

Laws cannot check if an object of any type is contained in a cell, only apply effect though an action /interaction from a player.

F.ex you cant tax based on how many bison mounts a player owns but you could based on how many he has plased down in total.

Same for conditions, you cannot explicit request how many objects has been placed down in x district but you can check if he places one down.
@WeaselDog

**Comment by Deantwo** [comment]
> Not entirely sure what would be the issue?
> 
> Laws cannot check if an object of any type is contained in a cell, only apply effect though an action /interaction from a player.
> 
> F.ex you cant tax based on how many bison mounts a player owns but you could based on how many he has plased down in total.
> 
> Same for conditions, you cannot explicit request how many objects has been placed down in x district but you can check if he places one down.

This was a question and confusion about how and when an action is counted as being in a district or on a claim. It has nothing to do with counting.

The question is/was about how action stats are saved in the database, and rather the "in district" and "on property" check is done when the action stat is recorded or when the action stats are queried.

If I place a bison mount on my property, then place another bison mount not on my property, and then claim the land that second bison mount was on. When I do a database query for how many bison mounts i have placed on my property, will the result be 1 or 2? That is the confusion.
